### PR TITLE
Air vent and gas pump improvements (and why we need them)

### PR DESCRIPTION
**Pull request by brndd**
[bug] [exploitable] [tweak] [OH_GOD_IT'S_LOOSE]

## How to remove all air on the station in less than 30 seconds

1) Go to the Atmospherics Automations Console
2) Import this script to it one line at a time, clicking on "Run Once" for each line: https://gist.github.com/brndd/9bb96ad2b088a729885a8114ad73c898
3) Wait about two ticks (4 seconds).
4) All the air from every room with a unary vent pump in it will now be gone (it's sucked back into distro, putting distro at about 9000 kPa). Yes, all of the station except for maint and other areas with no vents in them is now a perfect vacuum.
5) For extra credits, connect distro to waste and add a supercooling waste dump setup outside to make it very very annoying to repressurize the station.

Video: https://streamable.com/4q0kb

This honestly has so much round gunking potential compared to the effort it takes to set it up, even by atmos standards, that I'm a bit queasy about sharing it, but oh well.

## Why this works

Because our vent code (and pipe code in general) is fucking terrible and disabling pressure checks on vents them makes them fucking insane.

## What is being done about it?

This PR is what's being done about it. I ported over a bunch of pipe code fixes from Bay (don't worry, they're from before they went AGPLv3, so no licensing woes here) that fix this.

- Adds a total volume variable to pipenet datums. This is cached when the pipenet is updated, so it *shouldn't* degrade performance in any measurable way, but I couldn't be arsed to actually profile it.
- By using this total volume variable (plus some sanity checks on sucking air out of tiles), vents no longer have ridicilous turbogunking potential.

In practice this makes vents set to purge (T/N: purge means suck) drain air *roughly* as fast as scrubbers set to syphon. They drain air faster (MUCH faster) if there's lots and lots of pressure in the tile they're draining from (such as in the gas miner chambers, where they will pressurize the output pipe to several tens of thousands of kPa in a matter of seconds; they always did this by the way). I'm not a hydraulic engineer so I honestly don't have a perfect understanding of the numbers here, but they seem to work well enough in practice.

I also ported over a little optimization bit for vents. They now "hibernate" for 5-10 seconds if the room is stable (ie. not missing any gas), which avoids lots of pointless vent calculations being ran every tick.

## Oh, yeah, and gas pumps

The volume variable above allowed me to port over improvements to gas pumps (the pressure-based ones) from Bay too. They're no longer fucking garbage. In fact, they're now something on the order of a thousand times better than they used to be, depending on the situation. The pump in the setup below goes from transferring about 7.5 moles per tick to transferring about 7500 moles per tick (in the beginning—this slows down as distro fills up). This allows it to pressurize the distro loop to 103 kPa in about 1.5 to 2 minutes.

This is because gas pumps used to only check the volume of the pipe they're pumping into, rather than the whole pipenet, so they only moved enough gas to bring the immediate next pipe over to the target pressure. This is why they work reasonably well when filling canisters using connector ports with a gas pump right next to the connector, but do absolutely jack shit when hooked up to the several thousand pipes large main distro loop.

## Testing

I tested this in regular scenarios. I didn't test it with hardcore pipe autism. There may be unforeseen consequences for that, but since hardcore pipe autism usually relies on exploits anyway I don't think that should be a concern. If anything, it'll give hardcore pipe autists something new to figure out.

:cl:
 * bugfix: Vents can no longer drain the entire station of all its air in less than ten seconds.
 * bugfix: Passive vents now function like one would expect them to, balancing the external and internal pressure.
 * tweak: Gas pump formula was fixed, making them about 3 orders of magnitude better under the conditions where they used to be garbage (such as the distro gas pump in atmospherics). This basically gives them parity with volume pumps under optimal conditions.
 * tweak: Passive gates now function like one-way check valves, letting gas pass from input to output if input is higher pressure, but never from output to input. In practice you can no longer define their output pressure, but they also have no upper output pressure bound.

### CODE REVIEW

**Comment by DDihanov** [comment]
He did it, the madman actually did it!

**Comment by d3athrow** [comment]
If you're porting things from bay make sure you've gotten permission from them to license it as gplv3
agpl is like aids

**Comment by brndd** [comment]
@d3athrow these are from before they went AGPL.

**Comment by d3athrow** [comment]
Good shit then

**Comment by brndd** [comment]
I forgot about passive vents. Did you know that currently they're one-way volume pumps with no pressure limit?

Now they function similarly to regular vents, but aim to stabilize the external and internal pressure. They suck if external pressure is higher and blow if internal pressure is higher.

Oh yeah and I increased their volume to 500 from 250. Reason: vents used to be 200 volume and were buffed to 400 at some point. Passive vents were forgotten about, which left their volume as worse than that of a regular vent. You can weld them now too.

**Comment by brndd** [comment]
I changed passive gates to behave like one-way check valves. Their old behaviour was really kind of useless: they were just gas pumps that wouldn't actually pump. I don't think any of our atmos machines actually use power to do work (rather they just use it to keep the lights on), so they weren't terribly useful for anything.

Now they stabilize output with input, but only if input pressure is higher than output. If output is higher, they do nothing. In other words, they only let gas through one way. This means that they now also have no upper pressure bound. In practice this is the only difference to their old behaviour: you can no longer change the pressure they output, but the output pressure is now unbound.

**Comment by brndd** [comment]
Ugh, that created mapping gas. Replacing the passive gates on the affected maps (ie. maps that used passive gates with a specific output pressure) with gas pumps.

**Comment by brndd** [comment]
Alright, and now they work without power (like they used to claim they would), no longer have a pointless HTML interface with just an on/off button (they're now toggled on and off like valves just by clicking on them), and can be controlled remotely via signals from the AAC or other sources.

File: code/modules/atmos_automation/implementation/passive_gates.dm [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/datum/automation/set_passive_gate_state
+	name = "Passive Gate: Set Open/Closed"
+	var/gate=null
+	var/state=0
+
+	Export()
```
Comment (Kurfursten) [inline code/modules/atmos_automation/implementation/passive_gates.dm:6]: I know all the other automations use it but I don't feel like we should be continuing to promote relative pathing

Comment (PaulRitter) [reply code/modules/atmos_automation/implementation/passive_gates.dm:6]: @brndd you should be ashamed of yourself

Comment (brndd) [reply code/modules/atmos_automation/implementation/passive_gates.dm:6]: It wasn't removed from the other automations so I wanted to be consistent. I don't think it's all that bad here, this is almost a singleton datum and you should never see its procs called directly anywhere in the code.

**Comment by SonixApache** [comment]
Saw this abused today, very cute.

**Comment by brndd** [comment]
@Exxion pointed out the temperature calculation was backwards, so that's fixed now. With his help I also got my head around all this stuff better, so I actually kind of know what this stuff does now, hah hah.

Basically the pump change means that at their fastest they're now as fast as volume pumps. They'll never exceed the maximum speed of volume pumps (100 litres per second, despite the interface claiming 200), and their speed will gradually reduce as they get closer to their target pressure (due to some formula inaccuracies or something, I don't really know). They let you accurately control the target pressure, but on the other hand have a lower pressure bound than volume pumps (4500 for gas pumps and 9000 for volume pumps).

Passive gates are also at most as fast as volume pumps. I think I'll change this at some point, but it'll be after this PR.

File: code/modules/atmos_automation/implementation/passive_gates.dm (hunk 1 above)
Comment (Kurfursten) [reply code/modules/atmos_automation/implementation/passive_gates.dm:6]: I'm gonna let it slide this time but this is a bad practice. We shouldn't be coding for consistency with oldcode if oldcode isn't up to modern standards. Hopefully someone down the line fixes all of them, yours included.

**Pull request merged**